# Held Items: Stage 2 (Macho Brace)

This notebook is intentionally a **spec + checkpoint** surface for Stage 2.

Stage 1 math lives in `ev_core.py` and should be imported here rather than reimplemented.

## Scope

- Model Macho Brace as a multiplier on EV yield from a battle.
- Compare battles needed with vs without the item for the same trainee + opponent(s).
- Implementation belongs in a dedicated module once the behavior is locked (see `docs/workflow.md`).

## Imports

We import Stage 1 primitives from `ev_core.py`.

In [6]:
import math
import os
import sys

_REPO_ROOT = os.path.abspath(os.path.join(os.getcwd()))
if _REPO_ROOT not in sys.path:
    sys.path.insert(0, _REPO_ROOT)

from ev_core import battles_needed, ev_from_battle

## Base Inputs (same as Stage 1)

In [7]:
# TODO: Create a file with some example trainee + opponents
trainee = {
    "species": "Magikarp",
    "target_stat": "speed",
    "current_evs": 0,
    "target_evs": 252,
}

opponents = [
    {
        "species": "Geodude",
        "ev_yield_stat": "defense",
        "ev_yield": 1,
    },
    {
        "species": "Graveler",
        "ev_yield_stat": "defense",
        "ev_yield": 2,
    },
]

## Stage 1 Baseline Checkpoints

These should match `ev_basics.ipynb` after Stage 1 is complete.

In [8]:
print("Geodude EV gain (baseline):", ev_from_battle(opponents[0]))
print("Geodude battles (baseline):", battles_needed(trainee, opponents[0]))

print("Graveler EV gain (baseline):", ev_from_battle(opponents[1]))
print("Graveler battles (baseline):", battles_needed(trainee, opponents[1]))

Geodude EV gain (baseline): 1
Geodude battles (baseline): 252
Graveler EV gain (baseline): 2
Graveler battles (baseline): 126


## Stage 2 Design: Macho Brace

Let:

- `B` be battles needed
- `R` be remaining EVs needed for the trainee stat
- `Y` be total EV yield from the battle opponents list
- `M` be the Macho Brace multiplier (commonly `2`)

Baseline (Stage 1):

$$
B = \left\lceil \frac{R}{Y} \right\rceil
$$

With Macho Brace multiplying EV yield:

$$
B_{\text{macho}} = \left\lceil \frac{R}{M \cdot Y} \right\rceil
$$

Expected checkpoint for Geodude (`Y=1`, `M=2`, `R=252`):

$$
B_{\text{macho}} = \left\lceil \frac{252}{2} \right\rceil = 126
$$

In [ ]:
# Simulate a trainee with a held item
# TODO: trainee object should have a held item

macho_brace = {
    "item": "Macho Brace",
    "ev_multiplier": 2,
    "speed_multiplier": 0.5, # future operation-cost consideration; not used for EV math
}

held_item = macho_brace;

In [10]:
# Simulate battle with a held item
# TODO: these functions seem like they should be combined

# TODO: maybe rename the ev_from_battle function to ev_from_battle_baseline

def ev_from_battle_with_held_item(opponents, held_item):
    return ev_from_battle(opponents) * held_item["ev_multiplier"]

# TODO: maybe rename the battles_needed function to battles_needed_baseline

def battles_needed_with_held_item(trainee, opponents, held_item):
    remaining_evs_needed = trainee["target_evs"] - trainee["current_evs"]
    ev_gain = ev_from_battle_with_held_item(opponents, held_item)
    return math.ceil(remaining_evs_needed / ev_gain)

In [11]:
# Run the  held item battle

held_item_battle = ev_from_battle_with_held_item(opponents, held_item)
held_item_battles = battles_needed_with_held_item(trainee, opponents, held_item)

print(f"Held item battle EV gain: {held_item_battle}")
print(f"Held item battles needed: {held_item_battles}")

Held item battle EV gain: 6
Held item battles needed: 42


## Stage 2 Implementation Plan (no code yet)

Per the workflow, Stage 2 should add **new concrete functions** rather than rewriting Stage 1 APIs.

Sketch names only (final names TBD):

- `ev_from_battle_macho(opponents, macho_multiplier: int) -> int`
- `battles_needed_macho(trainee, opponents, macho_multiplier: int) -> int`

Both should be implemented in a module (for example `ev_items.py`) and covered by tests, then imported here for demos.

## Next Steps

See `docs/roadmap.md` for Stage 2 sequencing and `docs/workflow.md` for the extract + test + archive loop.